pkg/function/backup_data_using_kopia_server.go

```diff
@@ -0,0 +1,245 @@
+	}
+
+	var logSize, phySize int64
+	if snapInfo.Stats != nil {
+		stats := snapInfo.Stats
+		logSize = stats.SizeHashedB + stats.SizeCachedB
+		phySize = stats.SizeUploadedB
+	}
+
+	output := map[string]any{
+		BackupDataOutputBackupID:           snapInfo.SnapshotID,
+		BackupDataOutputBackupSize:         humanize.Bytes(uint64(logSize)),
+		BackupDataOutputBackupPhysicalSize: humanize.Bytes(uint64(phySize)),
+	}
+	return output, nil
+}
+
+func backupDataUsingKopiaServer(
+	cli kubernetes.Interface,
+	container,
+	hostname,
+	includePath,
+	namespace,
+	pod,
+	serverAddress,
+	fingerprint,
+	username,
+	userPassphrase string,
+	tags []string,
+) (info *kopiacmd.SnapshotCreateInfo, err error) {
+	contentCacheMB, metadataCacheMB := kopiacmd.GetCacheSizeSettingsForSnapshot()
+	configFile, logDirectory := kankopia.GetCustomConfigFileAndLogDirectory(hostname)
+
+	cmd := kopiacmd.RepositoryConnectServerCommand(kopiacmd.RepositoryServerCommandArgs{
+		UserPassword:    userPassphrase,
+		ConfigFilePath:  configFile,
+		LogDirectory:    logDirectory,
+		CacheDirectory:  kopiacmd.DefaultCacheDirectory,
+		Hostname:        hostname,
+		ServerURL:       serverAddress,
+		Fingerprint:     fingerprint,
+		Username:        username,
+		ContentCacheMB:  contentCacheMB,
+		MetadataCacheMB: metadataCacheMB,
+	})
+
+	stdout, stderr, err := kube.Exec(cli, namespace, pod, container, cmd, nil)
+	format.Log(pod, container, stdout)
+	format.Log(pod, container, stderr)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to connect to Kopia API server")
+	}
+
+	cmd = kopiacmd.SnapshotCreate(
+		kopiacmd.SnapshotCreateCommandArgs{
+			PathToBackup: includePath,
+			CommandArgs: &kopiacmd.CommandArgs{
+				RepoPassword:   "",
+				ConfigFilePath: configFile,
+				LogDirectory:   logDirectory,
+			},
+			Tags:                   tags,
+			ProgressUpdateInterval: 0,
+		})
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to generate snapshot create command")
+	}
+	stdout, stderr, err = kube.Exec(cli, namespace, pod, container, cmd, nil)
+	format.Log(pod, container, stdout)
+	format.Log(pod, container, stderr)
+
+	message := "Failed to create and upload backup"
+	if err != nil {
+		if strings.Contains(err.Error(), kerrors.ErrCodeOutOfMemoryStr) {
+			message = message + ": " + kerrors.ErrOutOfMemoryStr
+		}
+		return nil, errors.Wrap(err, message)
+	}
+	// Parse logs and return snapshot IDs and stats
+	return kopiacmd.ParseSnapshotCreateOutput(stdout, stderr)
+}
+
+func getHostNameAndUserPassPhraseFromRepoServer(userCreds string) (string, string, error) {
+	var userAccessMap map[string]string
+	if err := json.Unmarshal([]byte(userCreds), &userAccessMap); err != nil {
+		return "", "", errors.Wrap(err, "Failed to unmarshal User Credentials Data")
+	}
+
+	var userPassPhrase string
+	var hostName string
+	for key, val := range userAccessMap {
+		hostName = key
+		userPassPhrase = val
+	}
+
+	decodedUserPassPhrase, _ := base64.StdEncoding.DecodeString(userPassPhrase)
```

Comment:
does this return an error? Let's  not ignore that.

Reply:
Sure !